### DISCUSSION TITLE
OICD Problem with Apache Reverse Proxy

### DISCUSSION THREAD

**Discussion by dershodan** [outline/outline#7584] [Self Hosting] [answered]
I am trying to get outline running on our premises on a DMZ docker server.

TLS Termination and reverse proxying is handled by an **Apache 2.4** instance. The OICD Identity provider is **Keycloak**.

docker.env:
```
URL=https://mydomain.com
PORT=3000
```

Since communication between Apache and Outline is http i also set
`FORCE_HTTPS=false`

When the OICD redirect chain starts, outline tries to set a cookie in the 302 response headers (setting cookie and redirecting to keycloak).
Inspecting this in my browser shows the following Header:

`Set-Cookie: state=10.30.3.155 0b496b4b0dad54a4 web; path=/; expires=Wed, 11 Sep 2024 13:00:51 GMT; domain=10.30.3.155; httponly`
**Note**: 10.30.3.155 is actually the docker HOST ip - I do not know how the outline container knows about this IP at all.

This header is apparently not accepted by my browser, because after logging in in keycloak and getting redirected back to outline
(`auth/oidc.callback?state=<string>&session_state=<some 8-4-4-12 string>&iss=<keycloak domain>/realms/master&code=<some long string>`) i can see the request headers only contain this cookie header:

`Cookie: postLoginRedirectPath=%2Fdocs; XSRF-TOKEN=<some string>`

Outline logs the classic error: `{"error":"State not return in OAuth flow","level":"error","message":"Error during authentication", .....` which apparently is raised because the cookie is missing (from what i could learn online).

Here is my Apache Virtualhost:
```
<VirtualHost *:443>
      ServerName mydomain.com

      SSLEngine On
      SSLProtocol all -SSLv2 -SSLv3 -TLSv1 -TLSv1.1
      SSLHonorCipherOrder on
      SSLCipherSuite HIGH:!aNULL:!MD5:!3DES
      SSLCertificateFile <certificate path>
      SSLCertificateKeyFile <certificate path>
      SSLCACertificateFile <certificate path>

      SSLProxyEngine On
      ProxyPass / http://10.30.3.155:3000/
      ProxyPassReverse / http://10.30.3.155:3000/
</VirtualHost>
```

I am fairly certain, that I need to do more in the vhost, like add `X-Forwarded-Host` Headers and/or `Header always edit Set-Cookie .. ` but i tried any number of different variations - none seem to work.

So here I am humbly asking for your help so I no longer need to guess possible solutions.

**Comment by IamTaoChen** [comment, marked as answer]
```conf

<VirtualHost *:443>
    ServerName outline.example.com
    SSLEngine on

    # configure for ssl, this is global configuration.
    SSLCertificateFile /etc/apache2/certs/cert.pem
    SSLCertificateKeyFile /etc/apache2/certs/piavate.key

    # Websocket
    RewriteEngine On
    RewriteCond %{HTTP:Connection} Upgrade [NC]
    RewriteCond %{HTTP:Upgrade} websocket [NC]
    RewriteRule ^/?(.*) ws://outline.proxy:3000/$1  [P,L]
   
   # Proxy
    ProxyPreserveHost On
    <Location />
        RequestHeader set X_FORWARDED_PROTO 'https'
        RequestHeader set X-Forwarded-Ssl on
        RequestHeader set X-Forwarded-Port "443"
        ProxyPass http://outline:3000/
        ProxyPassReverse http://outline:3000/
    </Location> 

    # Log  
    ErrorLog ${APACHE_LOG_DIR}/note_error.log
    # CustomLog ${APACHE_LOG_DIR}/note_access.log combined 
</VirtualHost>
```

This is my apache2 conf

**Reply by dershodan** [reply]
Thank you so much, that solved it!
